File: articles/prenatal-development-the-earliest-foundations-of-life/go/prenatal_development_summary.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
)

func col(header []string, name string) int {
	for i, h := range header {
		if h == name {
			return i
		}
	}
	return -1
}

func main() {
	if len(os.Args) < 2 {
		log.Fatal("Usage: go run prenatal_development_summary.go data/prenatal_development_foundations_panel.csv")
	}
	f, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	rows, err := csv.NewReader(f).ReadAll()
	if err != nil {
		log.Fatal(err)
	}
	if len(rows) < 2 {
		log.Fatal("CSV has no data rows")
	}

	header := rows[0]
	outcomeIdx := col(header, "early_outcome")
	careIdx := col(header, "effective_care")
	riskIdx := col(header, "developmental_risk")
	gestationIdx := col(header, "gestational_weeks")
	healthIdx := col(header, "maternal_health")
	if outcomeIdx < 0 || careIdx < 0 || riskIdx < 0 || gestationIdx < 0 || healthIdx < 0 {
		log.Fatal("Required columns missing")
	}

	count := 0
	outcomeSum, careSum, riskSum, gestationSum, healthSum := 0.0, 0.0, 0.0, 0.0, 0.0

	for _, row := range rows[1:] {
		outcome, e1 := strconv.ParseFloat(row[outcomeIdx], 64)
		care, e2 := strconv.ParseFloat(row[careIdx], 64)
		risk, e3 := strconv.ParseFloat(row[riskIdx], 64)
		gestation, e4 := strconv.ParseFloat(row[gestationIdx], 64)
		health, e5 := strconv.ParseFloat(row[healthIdx], 64)
		if e1 != nil || e2 != nil || e3 != nil || e4 != nil || e5 != nil {
			continue
		}
		count++
		outcomeSum += outcome
		careSum += care
		riskSum += risk
		gestationSum += gestation
		healthSum += health
	}

	fmt.Printf("Rows analyzed: %d\n", count)
	fmt.Printf("Mean early_outcome: %.4f\n", outcomeSum/float64(count))
	fmt.Printf("Mean effective_care: %.4f\n", careSum/float64(count))
	fmt.Printf("Mean developmental_risk: %.4f\n", riskSum/float64(count))
	fmt.Printf("Mean gestational_weeks: %.4f\n", gestationSum/float64(count))
	fmt.Printf("Mean maternal_health: %.4f\n", healthSum/float64(count))
}

File: articles/prenatal-development-the-earliest-foundations-of-life/cpp/prenatal_development_summary.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

std::vector<std::string> split_csv_line(const std::string& line) {
    std::vector<std::string> values;
    std::stringstream ss(line);
    std::string item;
    while (std::getline(ss, item, ',')) values.push_back(item);
    return values;
}

int find_column(const std::vector<std::string>& header, const std::string& name) {
    for (size_t i = 0; i < header.size(); ++i) {
        if (header[i] == name) return static_cast<int>(i);
    }
    return -1;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: prenatal_development_summary data/prenatal_development_foundations_panel.csv\n";
        return 1;
    }

    std::ifstream file(argv[1]);
    if (!file.is_open()) {
        std::cerr << "Unable to open CSV file\n";
        return 1;
    }

    std::string line;
    std::getline(file, line);
    auto header = split_csv_line(line);

    int outcome_col = find_column(header, "early_outcome");
    int care_col = find_column(header, "effective_care");
    int risk_col = find_column(header, "developmental_risk");
    int gestation_col = find_column(header, "gestational_weeks");
    int health_col = find_column(header, "maternal_health");

    long count = 0;
    double outcome_sum = 0.0, care_sum = 0.0, risk_sum = 0.0, gestation_sum = 0.0, health_sum = 0.0;

    while (std::getline(file, line)) {
        auto values = split_csv_line(line);
        if ((int)values.size() <= outcome_col || (int)values.size() <= care_col || (int)values.size() <= risk_col || (int)values.size() <= gestation_col || (int)values.size() <= health_col) continue;
        outcome_sum += std::stod(values[outcome_col]);
        care_sum += std::stod(values[care_col]);
        risk_sum += std::stod(values[risk_col]);
        gestation_sum += std::stod(values[gestation_col]);
        health_sum += std::stod(values[health_col]);
        count++;
    }

    std::cout << "Rows analyzed: " << count << "\n";
    std::cout << "Mean early_outcome: " << outcome_sum / count << "\n";
    std::cout << "Mean effective_care: " << care_sum / count << "\n";
    std::cout << "Mean developmental_risk: " << risk_sum / count << "\n";
    std::cout << "Mean gestational_weeks: " << gestation_sum / count << "\n";
    std::cout << "Mean maternal_health: " << health_sum / count << "\n";
    return 0;
}
